RE: Add the ability to reorder or change the shelves

xlisachan opened the pull request:
Fixes #14874 
** Currently failing Travis, as it is dependent on the migration file in #15204 **

This pull request adds the ability to assign a shelf module a position/enables its appearance on the homepage. It also allows the admin user to reorder the shelves.

The tests `discovery/tests/test_admin.py` and `shelves/tests/test_forms.py` have been updated.

Please review when you have a moment. Thanks!

eviljeff commented on src/olympia/shelves/admin.py:
```diff
@@ -1,9 +1,30 @@
+    list_display_links = ('shelf',)
+    list_editable = ('position', 'enabled')
+    actions = ['delete_selected']
+    ordering = [F('position').asc(nulls_last=True)]
```
Interesting, I've not seen `F` objects in ordering before!

eviljeff replied:
you might not need the `F` object now `position` doesn't support null/None.

eviljeff commented on src/olympia/shelves/admin.py:
```diff
@@ -1,9 +1,30 @@
+    ordering = [F('position').asc(nulls_last=True)]
+    form = ShelfManagementForm
+
+    def save_model(self, request, obj, form, change):
```
Mmm, I'm not a fan of defining this validation twice.  I suggest moving the `def clean` from the form to `ShelfManagement` model.  (You probably don't need `ShelfManagementForm` at that point either - just define `fields` on `ShelfManagementAdmin`)

eviljeff commented on src/olympia/shelves/models.py:
```diff
@@ -27,3 +27,15 @@ class Meta:
+class ShelfManagement(ModelBase):
+    shelf = models.OneToOneField(Shelf, on_delete=models.CASCADE)
+    enabled = models.BooleanField(default=False)
+    position = models.PositiveIntegerField(blank=True, null=True)
```
when is it necessary to have `position` is `None`?   Wouldn't you just delete the `ShelfManagement` object?

xlisachan replied:
Possibly. With the enabled field, the user has the option of keeping shelf/shelves without enabling them on the homepage. My concern was to catch the case where the admin is reordering enabled shelves and just forgets to set the position before saving.

eviljeff replied:
it doesn't default to `0`? (or some other non-None number)

xlisachan replied:
No, default appears to be None. Should I set it to 0?

eviljeff replied:
yes, i think so (it might happen automatically if `null=True` wasn't specified)

xlisachan replied:
It didn't happen automatically, but `def clean` in models would no longer be needed if default is set to `0`, correct?

eviljeff replied:
correct.  If the field didn't allow `None` as a value then it would trigger the field's validation instead.

eviljeff commented on src/olympia/shelves/models.py:
```diff
@@ -27,3 +27,15 @@ class Meta:
+    position = models.PositiveIntegerField(blank=True, null=True)
+
+    def __str__(self):
+        return f'{self.shelf}'
```
```suggestion
        return str(self.shelf)
```

eviljeff commented on src/olympia/shelves/tests/test_forms.py:
```diff
@@ -116,3 +117,32 @@ def test_clean_returns_empty(self):
+        with self.assertRaises(ValidationError) as exc:
+            form.clean()
+        assert exc.exception.message == (
+            u'Position field is required to enable shelf.')
```
No need for `u'...'` strings - we're way past python2 compatibility 😄  (they only still exist in the codebase because it will be a huge patch to remove them all)

xlisachan commented on src/olympia/discovery/admin.py:
```diff
@@ -133,6 +133,12 @@ def previews(self, obj):
+class HomepageShelves(ShelfManagement):
+    class Meta:
+        proxy = True
+        verbose_name_plural = 'homepage shelves'
```
"Sentence Case" for `verbose_name_plural`? It didn't seem so. [https://docs.djangoproject.com/en/3.0/ref/models/options/#verbose-name-plural](url)

xlisachan commented:
Thanks for the feedback! This latest push includes the following updates:
- Revised the `ShelfManagement` model to set the default of `position` to 0, which eliminates the need for `def clean`
- Removed the `u'...'` strings
- Updated the return of `def __str__` for the `ShelfManagement` model

Please review when you have a moment. Thanks!
